### PR TITLE
8313224: Avoid calling JavaThread::current() in MemAllocator::Allocation constructor

### PR DESCRIPTION
**Pull request by iklam**
Summary: Avoid calling `JavaThread::current()` since we already have a `Thread*` in `MemAllocator::_thread`.

Although the latter is declared to be a `Thead*` and not a `JavaThread*`, the calling context requires the current thread to be a `JavaThread`, and  `MemAllocator::_thread` was initialized to be  `Thread::current`, so I changed the code from 

```
_thread(JavaThread::current()),
```

to

```
assert(Thread::current()->is_Java_thread(), "must be used by JavaThreads only");
assert(Thread::current() == allocator._thread, "do not pass MemAllocator across threads");
_thread = JavaThread::cast(allocator._thread);
```

I also clean up a few other lines to limit the explicit use of `JavaThread*`.

This fix improved `java --version` by about 0.6% for my prototype of [JDK-8310823](https://bugs.openjdk.org/browse/JDK-8310823), which allocates ~24000 heap objects at VM start-up.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8313224](https://bugs.openjdk.org/browse/JDK-8313224): Avoid calling JavaThread::current() in MemAllocator::Allocation constructor (**Enhancement** - P4)

### Reviewers
 * [Thomas Schatzl](https://openjdk.org/census#tschatzl) (@tschatzl - **Reviewer**) ⚠️ Review applies to [da211ff9](https://git.openjdk.org/jdk/pull/15058/files/da211ff932cbc115dba150c0060eba93e16666fc)
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15058/head:pull/15058` \
`$ git checkout pull/15058`

Update a local copy of the PR: \
`$ git checkout pull/15058` \
`$ git pull https://git.openjdk.org/jdk.git pull/15058/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15058`

View PR using the GUI difftool: \
`$ git pr show -t 15058`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15058.diff">https://git.openjdk.org/jdk/pull/15058.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15058#issuecomment-1654565499)

### CODE REVIEW

File: src/hotspot/share/gc/shared/memAllocator.cpp [hunk 1]
```diff
@@ -95,7 +97,7 @@ class MemAllocator::Allocation::PreserveObj: StackObj {
   oop* const _obj_ptr;
 
 public:
-  PreserveObj(JavaThread* thread, oop* obj_ptr)
+  PreserveObj(Thread* thread, oop* obj_ptr)
```
Comment (dholmes-ora) [inline src/hotspot/share/gc/shared/memAllocator.cpp:100]: Why did you change this? Following through the usage call chain it is only used by the `_thread` from `Allocation` which has to be a `JavaThread`.

**Review by dholmes-ora — COMMENTED** [review]
Generally looks okay but one query below.

Thanks.

File: src/hotspot/share/gc/shared/memAllocator.cpp (hunk 1 above)
Comment (iklam) [reply src/hotspot/share/gc/shared/memAllocator.cpp:100]: I wanted to minimize the use of `JavaThread`. Most of the MemAllocator APIs take `Thread`, so any occurrence of `JavaThread` looks suspicious and may cause people to unwittingly cast from `Thread` to `JavaThread` in the future.

File: src/hotspot/share/gc/shared/memAllocator.cpp [hunk 2]
```diff
@@ -70,19 +70,21 @@ class MemAllocator::Allocation: StackObj {
...
   Allocation(const MemAllocator& allocator, oop* obj_ptr)
     : _allocator(allocator),
-      _thread(JavaThread::current()),
       _obj_ptr(obj_ptr),
       _overhead_limit_exceeded(false),
       _allocated_outside_tlab(false),
       _allocated_tlab_size(0),
       _tlab_end_reset_for_sample(false)
   {
+    assert(Thread::current()->is_Java_thread(), "must be used by JavaThreads only");
```
Comment (tschatzl) [inline src/hotspot/share/gc/shared/memAllocator.cpp:79]: This first assert duplicates the one done by `JavaThread::cast()` given the other assert just below.

File: src/hotspot/share/gc/shared/memAllocator.cpp [hunk 3]
```diff
@@ -70,19 +70,21 @@ class MemAllocator::Allocation: StackObj {
...
+    assert(Thread::current() == allocator._thread, "do not pass MemAllocator across threads");
+    _thread = JavaThread::cast(allocator._thread);
```
Comment (tschatzl) [inline src/hotspot/share/gc/shared/memAllocator.cpp:81]: The assignment could also be done in the initializer list as it seems simple enough.

**Review by tschatzl — APPROVED** [review]
lgtm, with some minor comments about superfluous asserts.

**Review by coleenp — COMMENTED** [review]
Well, it's good to remove some thread arguments to eliminate some source of confusion.

File: src/hotspot/share/gc/shared/memAllocator.cpp (hunk 1 above)
Comment (iklam) [reply src/hotspot/share/gc/shared/memAllocator.cpp:99]: Sometimes MemAllocator is used by non-Java threads (e.g., GC threads uses it to create filler objects).

Comment (iklam) [reply src/hotspot/share/gc/shared/memAllocator.cpp]: I reverted the change in `PreserveObj`, which is a private inner class of `MemAllocator::Allocation`, so the use of `JavaThread*` doesn't leak to the outside world.

**Review by coleenp — APPROVED** [review]
Looks good to me.

**Comment by iklam** [comment]
Thanks for the review @tschatzl  @coleenp 

/integrate
